File: week01/credit.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>

char *type(char *card, int *sum_ptr);

int main(void)
{
    char card[17];
    long number;

    number = get_long("Number: ");
    sprintf(card, "%ld", number);
    while (strlen(card) != 13 && strlen(card) != 15 && strlen(card) != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    int total_sum;
    char *name = type(card, &total_sum);

    if (total_sum % 10 == 0 && name != NULL)
    {
        printf("%s\n", name);
    }
    else
    {
        printf("INVALID\n");
    }
}

char *type(char *card, int *sum_ptr)
{
    int len = strlen(card);
    int sum = 0;
    char *name = NULL;
    int double_digit = 0;//利用0表示false
    for (int i = len - 1; i >= 0; i--)
    {
        int digit = card[i] - '0';//将ASCII字符转换为数字（0-9）

        if (double_digit)//判断真假
        {
            digit *= 2;
            sum += (digit > 9) ? (digit - 9) : digit;
        }
        else
        {
            sum += digit;
        }
        double_digit = !double_digit;//真假反转（保证双倍乘法只在偶数位进行）
    }
    *sum_ptr = sum;

    // 判断卡类型
    if (len == 15 && card[0] == '3' && (card[1] == '4' || card[1] == '7'))
    {
        name = "AMEX";
    }
    else if (len == 16 && card[0] == '5' && card[1] >= '1' && card[1] <= '5')
    {
        name = "MASTERCARD";
    }
    else if ((len == 13 || len == 16) && card[0] == '4')
    {
        name = "VISA";
    }

    return name;
}

File: week06/sentimental-credit/credit.py
def get_int_card():
    while True:
        n = input("Number: ")
        try:
            try_card = int(n)
            card = str(try_card)
            sum_card0 = 0
            sum_card1 = 0
            sum_card0 = sum(int(x) * 2 if int(x) * 2 < 10 else int(x) * 2 - 9 for x in card[-2::-2])#将字符串切片转为整形
            sum_card1 = sum(int(x) for x in card[-1::-2])
            sum_card =sum_card1 + sum_card0
            return card, sum_card
        except ValueError:
            print("Value error,please enter ture card numbers ")

def first_judge_card(card, sum_card):
    if sum_card % 10 != 0:
        return False
    if len(card) not in [13, 14, 15, 16]:
        return False
    return True

def type_card(card):
    length = len(card)
    start = card[0:2]

    # 按最严格条件优先匹配
    if length == 15 and start in ['34', '37']:
        return 'AMEX'
    elif length == 16:
        if 51 <= int(start) <= 55:
            return 'MASTERCARD'
    if (length == 13 or length == 16) and card[0] == '4':
        return 'VISA'
    return 'INVALID'


card_number, sum_number = get_int_card()
if first_judge_card(card_number, sum_number):
    name = type_card(card_number)
    print("card_name:",name)
else:
    print("INVALID")
